File: app/src/main/java/de/mchllngr/quickopen/service/NotificationTileService.java
package de.mchllngr.quickopen.service;

import android.annotation.TargetApi;
import android.app.Service;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.Icon;
import android.os.Build;
import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;

import com.f2prateek.rx.preferences.Preference;
import com.f2prateek.rx.preferences.RxSharedPreferences;

import androidx.preference.PreferenceManager;
import de.mchllngr.quickopen.R;
import rx.Subscription;

/**
 * {@link Service} for handling the {@link Tile} for enabling and disabling the {@link android.app.Notification} in Android Nougat and above.
 */
@TargetApi(Build.VERSION_CODES.N)
public class NotificationTileService extends TileService {

    /**
     * {@link Preference}-reference for easier usage of the saved value for notificationEnabled in the {@link RxSharedPreferences}.
     */
    private Preference<Boolean> notificationEnabledPref;
    /**
     * {@link Subscription} from subscribing to {@code notificationEnabledPref}.
     */
    private Subscription notificationEnabledSubscription;

    @Override
    public void onCreate() {
        super.onCreate();

        RxSharedPreferences rxSharedPreferences = RxSharedPreferences.create(PreferenceManager.getDefaultSharedPreferences(this));

        notificationEnabledPref = rxSharedPreferences.getBoolean(
                getString(R.string.pref_notification_enabled),
                Boolean.parseBoolean(getString(R.string.pref_notification_enabled_default_value))
        );
    }

    @Override
    public void onStartListening() {
        super.onStartListening();

        // subscribe to changes in notificationEnabledPref
        notificationEnabledSubscription = notificationEnabledPref
                .asObservable()
                .subscribe(notificationEnabled -> {
                    if (notificationEnabled)
                        updateTileState(Tile.STATE_ACTIVE);
                    else
                        updateTileState(Tile.STATE_INACTIVE);
                });
    }

    /**
     * Updates the state of the shown {@link Tile}.
     */
    private void updateTileState(int state) {
        Tile tile = getQsTile();
        if (tile != null) {
            tile.setState(state);
            Icon icon = tile.getIcon();
            switch (state) {
                case Tile.STATE_ACTIVE:
                    icon.setTint(Color.WHITE);
                    break;
                case Tile.STATE_INACTIVE:
                case Tile.STATE_UNAVAILABLE:
                default:
                    icon.setTint(Color.GRAY);
                    break;
            }
            tile.updateTile();
        }
    }

    @Override
    public void onClick() {
        super.onClick();

        int i = getQsTile().getState();
        if (i == Tile.STATE_ACTIVE) {
            notificationEnabledPref.set(false);
        } else if (i == Tile.STATE_INACTIVE) {
            notificationEnabledPref.set(true);
            startNotificationService();

        }
    }

    @Override
    public void onStopListening() {
        if (notificationEnabledSubscription != null && !notificationEnabledSubscription.isUnsubscribed())
            notificationEnabledSubscription.unsubscribe();

        super.onStopListening();
    }

    @Override
    public void onDestroy() {
        notificationEnabledSubscription = null;
        notificationEnabledPref = null;

        super.onDestroy();
    }

    /**
     * Starts the {@link NotificationService}.
     */
    private void startNotificationService() {
        startService(new Intent(this, NotificationService.class));
    }
}

File: app/src/main/java/de/mchllngr/quickero/service/NotificationTileService.kt
package de.mchllngr.quickero.service

import android.annotation.TargetApi
import android.app.Notification
import android.app.Service
import android.graphics.Color
import android.os.Build
import android.service.quicksettings.Tile
import android.service.quicksettings.TileService
import dagger.hilt.android.AndroidEntryPoint
import de.mchllngr.quickero.repository.notification.NotificationRepository
import de.mchllngr.quickero.util.notification.NotificationHelper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import javax.inject.Inject

/** [Service] for handling the [Tile] for enabling and disabling the [Notification] in Android Nougat and above. */
@TargetApi(Build.VERSION_CODES.N)
@AndroidEntryPoint
class NotificationTileService : TileService() {

    private val serviceJob = SupervisorJob()
    private val serviceScope = CoroutineScope(Dispatchers.Main + serviceJob)

    private var listeningJob: Job? = null

    @Inject lateinit var notificationRepository: NotificationRepository
    @Inject lateinit var notificationHelper: NotificationHelper

    override fun onDestroy() {
        serviceJob.cancel()
        super.onDestroy()
    }

    override fun onStartListening() {
        super.onStartListening()

        listeningJob?.cancel()

        val job = SupervisorJob()
        listeningJob = job

        serviceScope.launch(job) {
            notificationRepository.enabled.collect {
                val state = if (it) Tile.STATE_ACTIVE else Tile.STATE_INACTIVE
                updateTileState(state)
            }
        }
    }

    override fun onStopListening() {
        listeningJob?.cancel()
        listeningJob = null
        super.onStopListening()
    }

    private fun updateTileState(state: Int) {
        val tile = qsTile ?: return

        tile.state = state
        tile.icon?.apply {
            val tint = when (state) {
                Tile.STATE_ACTIVE -> Color.WHITE
                else -> Color.GRAY
            }
            setTint(tint)
        }

        tile.updateTile()
    }

    override fun onClick() {
        super.onClick()

        val job = listeningJob ?: return
        serviceScope.launch(job) {
            val currentState = qsTile.state == Tile.STATE_ACTIVE
            val nextState = !currentState
            notificationRepository.setEnabled(nextState)
            if (nextState) notificationHelper.startNotificationService()
        }
    }
}
